proxy/src/main/java/net/md_5/bungee/util/ChatComponentTransformer.java

```diff
@@ -0,0 +1,85 @@
+package net.md_5.bungee.util;
+
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.chat.ScoreComponent;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.score.Score;
+import net.md_5.bungee.api.score.Scoreboard;
+import org.apache.commons.lang.Validate;
+
+import java.util.regex.Pattern;
+
+/**
+ * This class transforms chat components by attempting to replace transformable fields with the appropriate value.<br>
+ * ScoreComponents are transformed by replacing their {@link ScoreComponent#getName()}} into the matching entity's name
+ * as well as replacing the {@link ScoreComponent#getValue()} with the matching value in the {@link net.md_5.bungee.api.score.Scoreboard}
+ * if and only if the {@link ScoreComponent#getValue()} is not present.
+ */
+public final class ChatComponentTransformer
+{
+    /**
+     * The Pattern to match entity selectors.
+     */
+    private static final Pattern SELECTOR_PATTERN = Pattern.compile( "^@([pares])(?:\\[([^ ]*)\\])?$" );
+
+    /**
+     * Transform a given component, and attempt to transform the transformable fields.<br>
+     * Entity selectors <b>cannot</b> be evaluated.
+     * @param component the component to transform
+     * @return the transformed component
+     * @throws IllegalArgumentException if an entity selector pattern is present
+     */
+    public BaseComponent transform(BaseComponent component, Scoreboard scoreboard, ProxiedPlayer player)
```

Comment:
This isn't right, it needs to recursively format an array of BaseComponents, and also be called from sendMessage

Reply:
Is each BaseComponent a message to be sent? Do you want it formatted as a String? How should the end result look?